docs/src/main/asciidoc/websockets-next-reference.adoc

```diff
@@ -655,18 +655,58 @@ public class Endpoint {
 
 `SecurityIdentity` is initially created during a secure HTTP upgrade and associated with the websocket connection.
 
-Currently, for an HTTP upgrade be secured, users must configure an HTTP policy protecting the HTTP upgrade path.
-For example, to secure the `open()` method in the above websocket endpoint, one can add the following authentication policy:
+NOTE: When OpenID Connect extension is used and token expires, Quarkus automatically closes connection.
 
-[source,properties]
+== Secure HTTP upgrade
+
+An HTTP upgrade is secured when standard security annotation is placed on an endpoint class or an HTTP Security policy is defined.
+The advantage of securing HTTP upgrade is less processing, the authorization is performed early and only once.
+You should always prefer HTTP upgrade security unless, like in th example above, you need to perform action on error.
+
+.Use standard security annotation to secure an HTTP upgrade
+[source, java]
 ----
-quarkus.http.auth.permission.secured.paths=/end
-quarkus.http.auth.permission.secured.policy=authenticated
+package io.quarkus.websockets.next.test.security;
+
+import io.quarkus.security.Authenticated;
+import jakarta.inject.Inject;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.websockets.next.OnOpen;
+import io.quarkus.websockets.next.OnTextMessage;
+import io.quarkus.websockets.next.WebSocket;
+
+@Authenticated <1>
+@WebSocket(path = "/end")
+public class Endpoint {
+
+    @Inject
+    SecurityIdentity currentIdentity;
+
+    @OnOpen
+    String open() {
+        return "ready";
+    }
+
+    @OnTextMessage
+    String echo(String message) {
+        return message;
+    }
+}
 ----
+<1> Initial HTTP handshake ends with the 401 status for anonymous users.
+You can also redirect the handshake request on authorization failure with the `quarkus.websockets-next.server.security.auth-failure-redirect-url` configuration property.
 
-Other options for securing HTTP upgrade requests, such as using the security annotations, will be explored in the future.
+IMPORTANT: HTTP upgrade is only secured when a security annotation is declared on an endpoint class next to the `@WebSocket` annotation.
+Placing a security annotation on an endpoint bean will not secure bean methods, only the HTTP upgrade.
```

Comment:
@michalvavrik So, in the example above, `open()` is secured and `echo` is not ? It would be unexpected IMHO.

Reply:
@michalvavrik Michal, these are low level technical details that HTTP upgrade happens before `onOpen`, to me, on `onOpen` actually represents a successful HTTP upgrade completion, even if it is not quite a correct picture. 

Let me clarify how it aligns with #40534.

So there we have an HTTP security policy securing an upgrade:
```
quarkus.http.auth.permission.secured.paths=/end
quarkus.http.auth.permission.secured.policy=authenticated
```

and then we have
```
@WebSocket(path = "/end")
public class Endpoint {
    @Inject
    SecurityIdentity currentIdentity;
    
    @RolesAllowed("admin")
    @OnTextMessage
    String echo(String message) {
        return message;
    }
}
```

so `@RolesAllowed` must be placed on the `echo` method to ensure the current security identity is present and matches the sec constraint before `echo` is called.

Let's say, with your PR, I only want that `echo` is accessed by the non-anonymous security identity.

Are you saying we must do:

```
@WebSocket(path = "/end")
@Authenticated
public class Endpoint {
    @Inject
    SecurityIdentity currentIdentity;
    
    @Authenticated
    @OnTextMessage
    String echo(String message) {
        return message;
    }
}
```

to have `echo` secured ?

Also, does 
```
@WebSocket(path = "/end")
@RolesAllowed("admin")
public class Endpoint {
 
}
```  

works for a secure HTTP upgrade ? And if yes, should users also duplicate @RolesAllowed("admin") on the `echo` method for the `echo` be secured ?

I guess, all I'm asking, why can't we have a single `@Authenticated` at the class level, and not require users duplicate across various methods like onOpen, onEcho, etc, if all users want is, for ex, an authenticated access ?

Reply:
> Are you saying we must do:
> 
> ```
> @WebSocket(path = "/end")
> @Authenticated
> public class Endpoint {
>     @Inject
>     SecurityIdentity currentIdentity;
>     
>     @Authenticated
>     @OnTextMessage
>     String echo(String message) {
>         return message;
>     }
> }
> ```
> 
> to have `echo` secured ?

Nope, `@Authenticated` declared on an endpoint means that no callback can be called for non-anonymous security identity. In other words, we secure the HTTP upgrade and if the check fails then no method declared on the endpoint is invoked.

Reply:
> Also, does
> 
> ```
> @WebSocket(path = "/end")
> @RolesAllowed("admin")
> public class Endpoint {
>  
> }
> ```
> works for a secure HTTP upgrade ? And if yes, should users also duplicate @RolesAllowed("admin") on the echo method for the echo be secured ?

It's the same principle, i.e. no need to duplicate the annotation.

Reply:
> @michalvavrik Michal, these are low level technical details that HTTP upgrade happens before onOpen, to me, on onOpen actually represents a successful HTTP upgrade completion, even if it is not quite a correct picture

I planned to do it as you described in the #40622 but @mkouba pointed out that is is not quite true. `onOpen` does not represent the HTTP upgrade. Main difference is that if `onOpen` fails, what happens on `UnhandledFailureStrategy`. But if you rejected HTTP upgrade, you cannot allow for connection not being closed.

> So there we have an HTTP security policy securing an upgrade FOLLOWED BY HTTP Permissions policy

current implementation is absolutely equal to these HTTP permissions in your example, but securing `onOpen` instead would do more processing.

> works for a secure HTTP upgrade ? And if yes, should users also duplicate @RolesAllowed("admin") on the echo method for the echo be secured ?

yes it does work, I have tests for everything. you can use `@RolesAllowed` for both HTTP Upgrade and for `onMessage` and latter is called on every single message.

> I guess, all I'm asking, why can't we have a single @Authenticated at the class level, and not require users duplicate across various methods like onOpen, onEcho, etc, if all users want is, for ex, an authenticated access ?

but we do not require that, it is just possibility that I wanted to test; it works without it and I tried to stress in the documentation that recommended way is to secure HTTP upgrade. maybe it's a documentation issue?


FYI: IMO this discussion is theoretical but I am happy to answer all the questions; _theoretical because IMHO noone will inject endpoint and use it as CDI bean, it doesn't make sense. If you want, I can add validation for injection points (detectable ones, not programmatical) and check there is no security annotation present, so there is no risk._

Reply:
Thanks @mkouba, OK, I guess I'm finding the `Placing a security annotation on an endpoint bean will not secure bean methods, only the HTTP upgrade.` message a bit confusing, but this message can be improved.

Reply:
Can you suggest a message @sberyozkin ? It's optional, but I'd appreciate it. Otherwise I'll try to rewrite it, but I don't know if it will fix your suggestion.

Reply:
@michalvavrik 

>IMHO noone will inject endpoint and use it as CDI bean

May be this is the source of the confusion on my behalf and simply dropping this message will help ? See, I'm not thinking as a user in terms of CDI beans, when I see an example like
```
@WebSocket(path = "/end")
@Authenticated
public class Endpoint {
    @Inject
    SecurityIdentity currentIdentity;
    
    @OnTextMessage
    String echo(String message) {
        return message;
    }
}
```

I expect that the callback `echo` and `onOpen` methods are secured, I don't want to inject this endpoint anywhere, it is like a JAX-RS resource, we don't inject it somewhere else...

Reply:
Which is why reading `Placing a security annotation on an endpoint bean will not secure bean methods, only the HTTP upgrade.` confuses me a lot

Reply:
Anyway, I don't consider it as a blocker and minor updates to docs can be backported later if agreed to, so @mkouba @gsmet please merge if all looks good to you otherwise

Reply:
I think small follow-up docs PR where we can iterate would be better than do it on this PR as CI takes longer and I think we need more than attempt. +1 to merge this